File: src/main/java/org/azamov/learnjakarta/task2/java/filter/SecurityFilter.java
package org.azamov.learnjakarta.task2.java.filter;

import javax.servlet.*;
import javax.servlet.annotation.WebFilter;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.io.IOException;
import java.util.List;

@WebFilter("/*")
public class SecurityFilter implements Filter {
    private final List<String> publicPaths = List.of("/login", "/register");

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws IOException, ServletException {
        HttpServletRequest req = (HttpServletRequest) request;
        HttpServletResponse resp = (HttpServletResponse) response;

        String path = req.getServletPath();
        if (publicPaths.contains(path)) {
            chain.doFilter(request, response);
            return;
        }

        HttpSession session = req.getSession(false);
        Object username = session != null ? session.getAttribute("username") : null;

        if (username != null) {
            chain.doFilter(request, response);
        } else {
            resp.sendRedirect("/login");
        }
    }
}

File: src/main/java/org/azamov/learnjakarta/task2/kt/filter/SecurityFilter.kt
package org.azamov.learnjakarta.task2.kt.filter

import javax.servlet.Filter
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.annotation.WebFilter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
@WebFilter("/*")
class SecurityFilter : Filter {
    private val publicPaths = listOf("/login", "/register")

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        val req = request as HttpServletRequest
        val resp = response as HttpServletResponse

        val path = req.servletPath
        if (publicPaths.any { path.equals(it) }) {
            chain.doFilter(request, response)
            return
        }
        val session = req.getSession(false)
        val username = session?.getAttribute("username")

        if (username != null) {
            chain.doFilter(request, response)
        } else {
            resp.sendRedirect("/login")
        }
    }
}